### PR TITLE
Address Race Conditions and Better Error Handling to fix Issue #256

### PR DESCRIPTION
**Pull request by mearns**
## Description

Better error handling to address what I think issue #256 is about. There were also a couple of antipatterns being used and improper use of promises that lead to race conditions:
    
1. Using `process.exit()`  directly is usually not advisable as it prevents things from shutting down nicely. It also means if someone is trying to use this as a library, which I think may be the case for issue #256, you will exit from their code as well (i.e., error handling becomes impossible).

2. Functions that caught errors were printing them directly to STDERR rather than rethrowing them or throwing a new error. Again, if someone is trying to use it as a library as in the issues, this will make it impossible for them to do error handling.

3. There were some places where errors were being caught and printed to STDERR, but not causing the the process to exit with a non-zero exit code. It's always advisable to exit with a non-zero exit code when there is a failure.
    
4. There were some places where promises were not being waited for, leading to race conditions.

## Checklist

Please review this checklist before submitting a pull request.

- [X] Code compiles correctly
- [X] All tests passing (`npm run test:all`)

### CODE REVIEW

**Comment by agnivade** [comment]
Thanks @mearns ! The PR seems to have the older commit which got merged. Let's remove that.

**Comment by mearns** [comment]
Sure thing @agnivade. Except...I'm not sure how to =/. The only way I can think of is to close this PR and create a new branch from master. Sound ok?

**Comment by agnivade** [comment]
You can do a `git rebase -i HEAD~2` and then remove the earlier commit. And then do a `git push -f origin issues/256`. That should remove the commit and for a force push and update the PR.

**Comment by agnivade** [comment]
Thanks! I'd like @vladimyr to take a look.

**Comment by mearns** [comment]
Thanks for the git-fu, I wasn't sure if rebase was appropriate after it had been pushed, but it looks like it worked well =)

**Comment by agnivade** [comment]
Apologies for the delay here @mearns. I will get someone to review this.

File: lib/tldr.js [hunk 1]
```diff
@@ -23,34 +23,33 @@ class Tldr {
 
   list(singleColumn) {
     let os = platform.getPreferredPlatformFolder(this.config);
-    index.commandsFor(os)
+    return index.commandsFor(os)
```
Comment (sbrl) [inline lib/tldr.js:26]: This could potentially be refactored to use async / await to reduce visual complexity

Comment (mearns) [reply lib/tldr.js:26]: Sure. I didn't see async/await in there so I didn't want to introduce it, but I do prefer using it so if you'd prefer that syntax I can update it.

Comment (agnivade) [reply lib/tldr.js:26]: Up to you @mearns. Since you have already spent your time on it, I don't want to redo it again.

File: lib/tldr.js [hunk 2]
```diff
@@ -59,138 +58,107 @@ class Tldr {
...
+          })
+            .then(() => {
+              return spinningPromise('Creating index...', () => {
+                return search.createIndex();
+              });
+            })
+            .then(() => {
+              // And then, try to check in cache again
+              return this.cache.getPage(command);
+            });
```
Comment (agnivade) [inline lib/tldr.js:141]: Let's unindent these to match the closing branch of the previous block.

**Review by agnivade — COMMENTED** [review]
Just one small comment. Looks good otherwise.

File: lib/tldr.js (hunk 1 above)
Comment (mearns) [reply lib/tldr.js:26]: Looks like the projects current configuration doesn't support async/await syntax; it comes up with linting errors. I tried a few changes to address it, including updating to eslint version 7 and setting`parserOptions.ecmaVersion` to 7, but neither of these worked.

More importantly, these keywords [aren't supported before Node v7.10](https://node.green/#ES2017-features-async-functions), so to add them would require dropping support for older versions of Node (currently specified in package.json:engines as `>=6.12.0`), or else adding a transpiler.

In light of this, I propose that adding async/await is out of scope for this change and is a decision that the project owners need to make. If the decision is made to add them, I'd be happy to assist with that, but I propose it should be in a separate change set (a dedicated PR and release).

File: lib/tldr.js (hunk 2 above)
Comment (mearns) [reply lib/tldr.js:141]: If I understood the request correctly, eslint complains about that. I deleted two leading spaces from lines 133 through 141 (inclusive), so it looks like this:

```javascript
        if (!content) {
          // If not found, try to update
          return spinningPromise('Page not found. Updating cache...', () => {
            return this.cache.update();
          })
          .then(() => {
            return spinningPromise('Creating index...', () => {
              return search.createIndex();
            });
          })
          .then(() => {
            // And then, try to check in cache again
            return this.cache.getPage(command);
          });
        }
        return content;
      })
```

This results in 8 eslint errors due to violating the "indent" rule.

Comment (agnivade) [reply lib/tldr.js:141]: Well, that's a bit weird. But thanks for checking. :+1:

File: lib/tldr.js (hunk 1 above)
Comment (agnivade) [reply lib/tldr.js:26]: Let's go with the existing code for now.

**Review by agnivade — APPROVED** [review]
Thanks!

**Pull request merged**